### PR TITLE
Fix model server fails to gracefully shutdown

### LINKED ISSUE #3516: hugging face demo does not work in my local environment
I followed the guide here https://github.com/kserve/kserve/tree/master/python/huggingfaceserver. I tried run hugging face server locally, but it doesn't work.

**My running process is as follows:**

First, create a conda environment:
```
conda create -n kserve python=3.10
```

Then, install the necessary packages:
```
C:\Users\myusername\AppData\Roaming\Python\Scripts\poetry install
```

After that, run this command
```
 python -m huggingfaceserver --model_id=bert-base-uncased --model_name=bert
```

**Error message**

```
(kserve) PS D:\GithubRepo\kserve\python\huggingfaceserver> python -m huggingfaceserver --model_id=bert-base-uncased --model_name=bert
D:\AppData\Miniconda3\envs\kserve\lib\site-packages\timing_asgi\utils.py:9: UserWarning: resource module not available for Windows, cpu time will be set to 0
  warnings.warn("resource module not available for Windows, cpu time will be set to 0")
INFO:kserve:successfully loaded tokenizer for task: 5
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
INFO:kserve:successfully loaded huggingface model from path bert-base-uncased
INFO:kserve:Registering model: bert
INFO:kserve:Setting max asyncio worker threads as 24
INFO:kserve:Starting uvicorn with 1 workers
INFO:kserve:Starting gRPC server on [::]:8081
Traceback (most recent call last):
  File "D:\AppData\Miniconda3\envs\kserve\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "D:\AppData\Miniconda3\envs\kserve\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "D:\GithubRepo\kserve\python\huggingfaceserver\huggingfaceserver\__main__.py", line 70, in <module>
    kserve.ModelServer().start([model] if model.ready else [])
  File "D:\GithubRepo\kserve\python\kserve\kserve\model_server.py", line 226, in start
    asyncio.run(servers_task())
  File "D:\AppData\Miniconda3\envs\kserve\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "D:\AppData\Miniconda3\envs\kserve\lib\asyncio\base_events.py", line 649, in run_until_complete
    return future.result()
  File "D:\GithubRepo\kserve\python\kserve\kserve\model_server.py", line 224, in servers_task
    await asyncio.gather(*servers)
  File "D:\GithubRepo\kserve\python\kserve\kserve\model_server.py", line 188, in serve
    loop.add_signal_handler(
  File "D:\AppData\Miniconda3\envs\kserve\lib\asyncio\events.py", line 553, in add_signal_handler
    raise NotImplementedError
NotImplementedError
Traceback (most recent call last):
  File "D:\AppData\Miniconda3\envs\kserve\lib\site-packages\grpc\aio\_server.py", line 185, in __del__
  File "src\python\grpcio\grpc\_cython\_cygrpc/aio/common.pyx.pxi", line 120, in grpc._cython.cygrpc.schedule_coro_threadsafe   
  File "src\python\grpcio\grpc\_cython\_cygrpc/aio/common.pyx.pxi", line 112, in grpc._cython.cygrpc.schedule_coro_threadsafe   
  File "D:\AppData\Miniconda3\envs\kserve\lib\asyncio\base_events.py", line 436, in create_task
  File "D:\AppData\Miniconda3\envs\kserve\lib\asyncio\base_events.py", line 515, in _check_closed
RuntimeError: Event loop is closed
sys:1: RuntimeWarning: coroutine 'AioServer.shutdown' was never awaited
```

**Environment:**

My OS is Windows. My poetry version is `Poetry (version 1.8.2)`.

### LINKED ISSUE #3882: An attempt to run microservice on Windows fails with NotImplementedException
/kind bug

Running microservice on Windows fails with NotImplementedException, because ProactorEventLoop.add_signal_handler(...) method is not implemented on Windows.

**What did you expect to happen:**

A trivial workaround exists involving the following minor code change in the ModelServe.serve() function defined in model_server.py:

``` Python
if sys.platform not in ["win32", "win64"]:
    sig_list = [signal.SIGINT, signal.SIGTERM, signal.SIGQUIT]
else:
    # add_signal_handler is not implemented on Windows!
    # sig_list = [signal.SIGINT, signal.SIGTERM] commented out
    sig_list = []  # replaced with empty list
for sig in sig_list:
    loop.add_signal_handler(
        sig, lambda s=sig: asyncio.create_task(self.stop(sig=s))
    )
```

**Anything else you would like to add:**

I understand that the kserve package is designed to be used primarily on Linux-based platforms; however, it can also be used on Windows with somewhat limited functionality, such as development, testing and prototyping.

**Environment:**

- KServe Version: 0.13.1
- OS (e.g. from `/etc/os-release`): Windows 11
- Python 3.11.9

### LINKED ISSUE #4115: Model server fails to gracefully down rest server
/kind bug

**What steps did you take and what happened:**

The server fails to gracefully shut down due to the changes in the signal handling of uvicron server https://github.com/encode/uvicorn/pull/1600.
```
$ python3 /custom_model/model.py

/home/ubuntu/Documents/Bloomberg/kserve/.venv/lib/python3.12/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/ubuntu/Documents/Bloomberg/kserve/.venv/lib/python3.12/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=AlexNet_Weights.IMAGENET1K_V1`. You can also use `weights=AlexNet_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
2024-12-11 18:18:10.974 521739 kserve INFO [model_server.py:register_model():389] Registering model: model
2024-12-11 18:18:10.974 521739 kserve INFO [model_server.py:start():323] Setting max asyncio worker threads as 12
2024-12-11 18:18:10.975 521739 kserve INFO [model_server.py:_serve_rest():269] Starting uvicorn with 1 workers
2024-12-11 18:18:11.015 uvicorn.error INFO:     Started server process [521739]
2024-12-11 18:18:11.015 uvicorn.error INFO:     Waiting for application startup.
2024-12-11 18:18:11.017 521739 kserve INFO [server.py:start():68] Starting gRPC server on [::]:8081
2024-12-11 18:18:11.017 uvicorn.error INFO:     Application startup complete.
2024-12-11 18:18:11.017 uvicorn.error INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
^C2024-12-11 18:18:12.716 521739 kserve INFO [model_server.py:stop():343] Stopping the model server
2024-12-11 18:18:12.716 521739 kserve INFO [model_server.py:stop():345] Stopping the rest server
2024-12-11 18:18:12.716 521739 kserve INFO [model_server.py:stop():348] Stopping the grpc server
2024-12-11 18:18:12.716 521739 kserve INFO [server.py:stop():74] Waiting for gRPC server shutdown
2024-12-11 18:18:12.717 521739 kserve INFO [server.py:stop():76] gRPC server shutdown complete
2024-12-11 18:18:12.724 uvicorn.error INFO:     Shutting down
2024-12-11 18:18:12.825 uvicorn.error INFO:     Waiting for application shutdown.
2024-12-11 18:18:12.825 uvicorn.error INFO:     Application shutdown complete.
2024-12-11 18:18:12.825 uvicorn.error INFO:     Finished server process [521739]
Traceback (most recent call last):
  File "/home/ubuntu/Documents/Bloomberg/kserve/python/custom_model/model.py", line 150, in <module>
    ModelServer().start([model])
  File "/home/ubuntu/Documents/Bloomberg/kserve/python/kserve/kserve/model_server.py", line 334, in start
    asyncio.run(servers_task())
  File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/ubuntu/Documents/Bloomberg/kserve/python/kserve/kserve/model_server.py", line 332, in servers_task
    await asyncio.gather(*servers)
  File "/home/ubuntu/Documents/Bloomberg/kserve/python/kserve/kserve/model_server.py", line 295, in _serve_rest
    await self._rest_server.run()
  File "/home/ubuntu/Documents/Bloomberg/kserve/python/kserve/kserve/protocol/rest/server.py", line 153, in run
    await self.server.serve()
  File "/home/ubuntu/Documents/Bloomberg/kserve/.venv/lib/python3.12/site-packages/uvicorn/server.py", line 68, in serve
    with self.capture_signals():
  File "/usr/lib/python3.12/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/ubuntu/Documents/Bloomberg/kserve/.venv/lib/python3.12/site-packages/uvicorn/server.py", line 328, in capture_signals
    signal.raise_signal(captured_signal)
TypeError: 'NoneType' object cannot be interpreted as an integer
```

**What did you expect to happen:**
The server should shut down gracefully

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]
Not Applicable

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: 0.14.0
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
- Fixes #4115 
- Fixes #3882
- Fixes #3516

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] Test A
Logs:
```
$ python3 custom_model/model.py

/home/ubuntu/Documents/Bloomberg/kserve/.venv/lib/python3.12/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/ubuntu/Documents/Bloomberg/kserve/.venv/lib/python3.12/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=AlexNet_Weights.IMAGENET1K_V1`. You can also use `weights=AlexNet_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
2024-12-11 18:34:56.181 528136 kserve INFO [model_server.py:register_model():368] Registering model: model
2024-12-11 18:34:56.181 528136 kserve INFO [model_server.py:setup_event_loop():286] Setting max asyncio worker threads as 12
2024-12-11 18:34:56.214 528136 kserve INFO [server.py:start():145] Starting uvicorn with 1 workers
2024-12-11 18:34:56.229 uvicorn.error INFO:     Started server process [528136]
2024-12-11 18:34:56.229 uvicorn.error INFO:     Waiting for application startup.
2024-12-11 18:34:56.231 528136 kserve INFO [server.py:start():68] Starting gRPC server with 4 workers
2024-12-11 18:34:56.231 528136 kserve INFO [server.py:start():69] Starting gRPC server on [::]:8081
2024-12-11 18:34:56.231 uvicorn.error INFO:     Application startup complete.
2024-12-11 18:34:56.232 uvicorn.error INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
^C2024-12-11 18:34:58.640 uvicorn.error INFO:     Shutting down
2024-12-11 18:34:58.740 uvicorn.error INFO:     Waiting for application shutdown.
2024-12-11 18:34:58.741 uvicorn.error INFO:     Application shutdown complete.
2024-12-11 18:34:58.741 uvicorn.error INFO:     Finished server process [528136]
2024-12-11 18:34:58.741 528136 kserve INFO [model_server.py:stop():325] Stopping the model server
2024-12-11 18:34:58.741 528136 kserve INFO [model_server.py:stop():327] Stopping the grpc server
2024-12-11 18:34:58.741 528136 kserve INFO [server.py:stop():75] Waiting for gRPC server shutdown
2024-12-11 18:34:58.742 528136 kserve INFO [server.py:stop():77] gRPC server shutdown complete
```

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

```release-note
Fix model server fails to gracefully shutdown
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: python/kserve/kserve/protocol/rest/server.py [hunk 1]
```diff
@@ -103,55 +111,36 @@ def create_application(self):
...
             # The asgi-logger settings don't set propagate to False,
             # so we get duplicates if we don't set it explicitly.
             logging.getLogger("access").propagate = False
 
-        self.server = _NoSignalUvicornServer(config=self.cfg)
-
-    async def run(self):
-        await self.server.serve()
-
-    async def stop(self, sig: Optional[int] = None):
```
Comment (sivanantha321) [inline python/kserve/kserve/protocol/rest/server.py:155]: Uvicron registers its [own signal handler and re raises the signal](https://github.com/encode/uvicorn/blob/bfa754e21e2cc1d5b0d7cabf24933a6c3afc315e/uvicorn/server.py#L315). So we don't have to handle the shutdown process.

File: python/kserve/kserve/model_server.py [hunk 2]
```diff
@@ -264,86 +253,77 @@ def __init__(
...
-                servers.append(self._grpc_server.start(self.max_threads))
-            await asyncio.gather(*servers)
+    def register_signal_handler(self):
+        if sys.platform == "win32":
+            sig_list = [signal.SIGINT, signal.SIGTERM, signal.SIGBREAK]
+        else:
+            sig_list = [signal.SIGINT, signal.SIGTERM, signal.SIGQUIT]
+
+        for sig in sig_list:
+            signal.signal(
```
Comment (sivanantha321) [inline python/kserve/kserve/model_server.py:300]: https://github.com/encode/uvicorn/pull/1600#issuecomment-1319933056. Using `loop.add_signal_handler` leads to the handler being called twice as uvicorn re raises the signal.

File: python/kserve/test/test_server.py [hunk 3]
```diff
@@ -1217,21 +1133,16 @@ def test_predict_ce_avro_binary(self, http_server_client):
...
-    @pytest.fixture(scope="class")
-    def app(self):
-        server = ModelServer(
-            registered_models=DummyModelRepository(test_load_success=True)
-        )
-        rest_server = RESTServer(
-            kserve_app, server.dataplane, server.model_repository_extension
+    @pytest_asyncio.fixture(scope="class", autouse=True)
+    def app(self, server):
+        mp = pytest.MonkeyPatch()
```
Comment (yuzisun) [inline python/kserve/test/test_server.py:1138]: instead of Monkey patch, we can create a different server fixture?

Comment (sivanantha321) [reply python/kserve/test/test_server.py:1138]: The fastapi app is global variable. Registering endpoints and middlewares again will lead to runtime errors.

**Pull request merged**